common/src/main/java/com/microsoft/identity/common/internal/platform/DevicePopManager.java

```diff
@@ -0,0 +1,834 @@
+            final RSASSASigner signer = new RSASSASigner(privateKey);
+
+            final SignedJWT signedJWT = new SignedJWT(
+                    new JWSHeader.Builder(JWSAlgorithm.RS256)
+                            .build(),
+                    claimsSet
+            );
+
+            signedJWT.sign(signer);
+
+            return signedJWT.serialize();
+        } catch (final NoSuchAlgorithmException e) {
+            exception = e;
+            errCode = NO_SUCH_ALGORITHM;
+        } catch (final KeyStoreException e) {
+            exception = e;
+            errCode = KEYSTORE_NOT_INITIALIZED;
+        } catch (final JOSEException e) {
+            exception = e;
+            errCode = JWT_SIGNING_FAILURE;
+        } catch (final UnrecoverableEntryException e) {
+            exception = e;
+            errCode = INVALID_PROTECTION_PARAMS;
+        }
+
+        final ClientException clientException = new ClientException(
+                errCode,
+                exception.getMessage(),
+                exception
+        );
+
+        Logger.error(
+                TAG,
+                clientException.getMessage(),
+                clientException
+        );
+
+        throw clientException;
+    }
+
+    //region Internal Functions
+
+    /**
+     * Generates a new RSA KeyPair of the specified lenth.
+     *
+     * @param ctx        The current application Context.
+     * @param minKeySize The minimum keysize to use.
+     * @return The newly generated RSA KeyPair.
+     * @throws UnsupportedOperationException
+     */
+    @SuppressLint("NewApi")
+    private KeyPair generateNewRsaKeyPair(@NonNull final Context ctx,
+                                          final int minKeySize)
+            throws UnsupportedOperationException, InvalidAlgorithmParameterException,
+            NoSuchAlgorithmException, NoSuchProviderException {
+        final int MAX_RETRIES = 4;
+
+        for (int ii = 0; ii < MAX_RETRIES; ii++) {
+            KeyPair kp;
+
+            try {
+                kp = generateNewKeyPair(ctx, true);
+            } catch (final StrongBoxUnavailableException e) {
+                Logger.error(
+                        TAG,
+                        "StrongBox unsupported - skipping hardware flags.",
+                        e
+                );
+
+                // Retry, but don't request StrongBox
+                kp = generateNewKeyPair(ctx, false);
+            }
+
+            // Due to a bug in some versions of Android, keySizes may not be exactly as specified
+            // To generate a 2048-bit key, two primes of length 1024 are multiplied -- this product
+            // may be 2047 in length in some cases which causes Nimbus to crash. To avoid this,
+            // check the keysize prior to returning the generated KeyPair.
+
+            // Since this seems to be nondeterministic in nature, attempt this a maximum of 4 times.
+            final int length = RSAKeyUtils.keyBitLength(kp.getPrivate());
```

Comment:
should we only loop here? (this line to be specific)

Reply:
Can't - `generateNewKeyPair` must be called between calls to `RSAKeyUtils.keyBitLength` in order to regenerate the private keys (which `keyBitLength` measures)